File: apps/etl-clojure/src/pulso/etl.clj
(ns pulso.etl
  (:require [clojure.tools.logging :as log]
            [pulso.db :as db]
            [pulso.xml.parser :as parser]
            [pulso.loader.profile :as profile]
            [pulso.loader.records :as records]
            [pulso.loader.workouts :as workouts]
            [pulso.loader.correlations :as correlations]
            [pulso.loader.activity :as activity]
            [pulso.loader.lookups :as lookups]))

(defn execute!
  "Runs the full ETL pipeline: parse XML → transform → load into PostgreSQL.

   Accepts an optional opts map:
     :on-element - callback fn called with element keyword after each element is processed.
                   When provided, the old log-every-100k progress reporting is disabled."
  ([ds xml-file batch-size]
   (execute! ds xml-file batch-size {}))
  ([ds xml-file batch-size opts]
   (log/info "Starting ETL pipeline" {:file xml-file :batch-size batch-size})
   (let [start        (System/currentTimeMillis)
         on-element   (:on-element opts)
         use-log-progress (nil? on-element)
         progress-every 100000]

     ;; Prepare: truncate for idempotency, reset caches
     (db/truncate-all! ds)
     (lookups/reset-caches!)
     (profile/reset-state!)

     ;; Initialize batchers
     (let [record-batchers  (records/make-batchers ds batch-size)
           workout-batchers (workouts/make-batchers ds batch-size)
           corr-batchers    (correlations/make-batchers ds batch-size)
           activity-batcher (activity/make-batcher ds batch-size)
           counters         (atom {:records 0 :workouts 0 :correlations 0 :activities 0})]

       ;; Stream and process
       (parser/parse-health-data xml-file
         (fn [element _locale]
           (case (:tag element)
             :ExportDate
             (do (profile/save-export-date! element)
                 (when on-element (on-element :ExportDate)))

             :Me
             (do (profile/save-profile! ds element _locale)
                 (when on-element (on-element :Me)))

             :Record
             (do (records/process! ds record-batchers element)
                 (let [n (swap! counters update :records inc)]
                   (when (and use-log-progress
                              (zero? (mod (:records n) progress-every)))
                     (log/info "Progress:" (:records n) "records processed")))
                 (when on-element (on-element :Record)))

             :Workout
             (do (workouts/process! ds workout-batchers element)
                 (swap! counters update :workouts inc)
                 (when on-element (on-element :Workout)))

             :Correlation
             (do (correlations/process! ds corr-batchers element)
                 (swap! counters update :correlations inc)
                 (when on-element (on-element :Correlation)))

             :ActivitySummary
             (do (activity/process! activity-batcher element)
                 (swap! counters update :activities inc)
                 (when on-element (on-element :ActivitySummary)))

             ;; Skip unknown elements
             (log/debug "Skipping element" {:tag (:tag element)}))))

       ;; Flush remaining partial batches
       (records/flush! record-batchers)
       (workouts/flush! workout-batchers)
       (correlations/flush! ds corr-batchers)
       ((:flush! activity-batcher))

       ;; Report
       (let [elapsed  (- (System/currentTimeMillis) start)
             counts   @counters]
         (log/info "ETL complete!" (assoc counts :elapsed-ms elapsed
                                                 :elapsed-min (format "%.1f" (/ elapsed 60000.0))))
         counts)))))

File: apps/etl-python/pulso/etl.py
import logging
import time

from pulso import db
from pulso.loader import activity, correlations, lookups, profile, records, workouts
from pulso.xml import parser

logger = logging.getLogger(__name__)

PROGRESS_EVERY = 100000


def execute(pool, xml_file, batch_size, on_element=None):
    """Runs the full ETL pipeline: parse XML -> transform -> load into
    PostgreSQL. When on_element is provided, the old log-every-100k
    progress reporting is disabled (the caller is driving its own UI)."""
    logger.info("Starting ETL pipeline file=%s batch_size=%s", xml_file, batch_size)
    start = time.time()
    use_log_progress = on_element is None

    db.truncate_all(pool)
    lookups.reset_caches()
    profile.reset_state()

    record_batchers = records.make_batchers(pool, batch_size)
    workout_batchers = workouts.make_batchers(pool, batch_size)
    corr_batchers = correlations.make_batchers(pool, batch_size)
    activity_batcher = activity.make_batcher(pool, batch_size)
    counters = {"records": 0, "workouts": 0, "correlations": 0, "activities": 0}

    def handler(element, locale):
        tag = element.tag
        if tag == "ExportDate":
            profile.save_export_date(element)
            if on_element:
                on_element("ExportDate")
        elif tag == "Me":
            profile.save_profile(pool, element, locale)
            if on_element:
                on_element("Me")
        elif tag == "Record":
            records.process(pool, record_batchers, element)
            counters["records"] += 1
            if use_log_progress and counters["records"] % PROGRESS_EVERY == 0:
                logger.info("Progress: %s records processed", counters["records"])
            if on_element:
                on_element("Record")
        elif tag == "Workout":
            workouts.process(pool, workout_batchers, element)
            counters["workouts"] += 1
            if on_element:
                on_element("Workout")
        elif tag == "Correlation":
            correlations.process(pool, corr_batchers, element)
            counters["correlations"] += 1
            if on_element:
                on_element("Correlation")
        elif tag == "ActivitySummary":
            activity.process(activity_batcher, element)
            counters["activities"] += 1
            if on_element:
                on_element("ActivitySummary")
        else:
            logger.debug("Skipping element tag=%s", tag)

    parser.parse_health_data(xml_file, handler)

    records.flush(record_batchers)
    workouts.flush(workout_batchers)
    correlations.flush(pool, corr_batchers)
    activity_batcher.flush()

    elapsed_ms = int((time.time() - start) * 1000)
    logger.info(
        "ETL complete! %s elapsed_ms=%s elapsed_min=%.1f",
        counters, elapsed_ms, elapsed_ms / 60000.0,
    )
    return counters
